Canister max. pressure capacity display fix

MadMedicineMan opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Atmos canisters now display their correct max. pressure capacity. Game code always expects kPa, code for SI-prefix-generation only expects the base unit (here: Pa), so the displayed value was always off by the factor 1000.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Did me a heckin confuse multiple times. Good for new atmos techs.

The improved solution has a separate proc to convert pressure, thus prevents the need for manual correction (x10e3) in code whenever pressure needs to be converted to readable text.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
fix: Canisters now display correct max. pressure
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MadMedicineMan committed:
Merge pull request #1 from tgstation/master

Fork update

MadMedicineMan committed:
Canisters display correct pressure capacity now

Canisters now display the correct pressure capacity. Previously due to a conversion bug, they displayed less by a factor of 1000

TetraK1 commented:
Would it not make slightly more sense to instead change the unit to kPa?  I can't think of another place where straight Pa is used.

MadMedicineMan commented:
> Would it not make slightly more sense to instead change the unit to kPa? I can't think of another place where straight Pa is used.

The unit everywhere is kPa, thats the problem. and the proc converting the Unit is used for every SI, like Amperes and Volt. I thought about changing the proc with this one exception but that seems a little like producing spaghetti.

Donkie commented:
All places should use siunit instead of hardcoding kPa display. Its dumb to see stuff like "0.01 kPa". tgui also properly uses siunits.

MadMedicineMan commented:
Alternative solution that also works and does - so far - not break any usage of the siunit proc:

`// The game code never uses Pa, but kPa, since 1 Pa is too small to reasonably handle
// thus, to ensure correct conversion from any kPa in game code, this value needs to be multiplied by 10e3 to get Pa
	if(unit == "Pa")
		value = value * 1000`

Which, imo, is pretty hacky, but is more futureproof. Because i doubt anyone will ever use this proc and **actually** use Pa instead of kPa.

MadMedicineMan committed:
Alternative canister solution

futureproof hack

MadMedicineMan committed:
further test fixes

MadMedicineMan committed:
Merge branch 'canister-fix-alternative' into canister-capacity-fix

Donkie commented:
I dont like that solution at all, stick with the previous one.

MadMedicineMan committed:
Cleanup

unnecessary annotation from previous solution

MadMedicineMan commented:
> I dont like that solution at all, stick with the previous one.

That was like 30 seconds too late. Damn.

MadMedicineMan committed:
Fixes oversight in Travis tests

Oversight in siunit.dm Line 8 was still written for the old proc.

MadMedicineMan committed:
More Travis fixes

Because I work to hasty. And I'm sorry. Also sorted the tests by Unit

Ghilker commented:
did you test this in game? i feel like this doesn't work

LemonInTheDark commented on code/__HELPERS/text.dm:
```diff
@@ -904,6 +904,13 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
 		return "0 [unit]"
 
+
+	// The game code never uses Pa, but kPa, since 1 Pa is too small to reasonably handle
+	// thus, to ensure correct conversion from any kPa in game code, this value needs to be multiplied by 10e3 to get Pa
+	if(unit == "Pa")
+		value = value * 1000
+
```
Does this match the other bits of code that use Pa here? I have a feeling this would break some assumptions.

Donkie replied:
This is a very ugly hack, make a new proc instead specifically made for si-printing pressure.

MadMedicineMan replied:
> Does this match the other bits of code that use Pa here? I have a feeling this would break some assumptions.

Which assumptions do you mean? As far as i can tell, if this proc is used it always uses "Pa" to reference pressure.

MadMedicineMan replied:
> This is a very ugly hack, make a new proc instead specifically made for si-printing pressure.

I feel like this is a little much for such a small proc, but the only way I think would make sense is to implement this for most of the standard SIs without tossing the previous coders work out the window (because their solution is rather nice). is basically a proc that is 

`proc/siunit_pressure(value_in_kpa, maxdecimals)`
`var/pressure_adj = value_in_kpa * 1000 //to adjust for using kPa instead of Pa`
`return siunit(pressure_adj, "Pa", maxdecimals)`

And then that for all other units that use the same proc in code (which isnt that much). if feels even more cumbersome. Or what did you have in mind?

MadMedicineMan replied:
and all the other SI dont really need this adjustment and can just call siunit(). Its really just kPa that is the troublemaker.

MadMedicineMan replied:
@Donkie Im toying around with your idea and tried the approach i detailed above. Should I implement it? so far it seems to work, and it does not break other SIs.

MadMedicineMan committed:
Cleaner version (Sep. proc for pressure conversion)

- Added separate proc for pressure conversion (siunit_pressure()) that calls siunit()
- All instances of siunit() calls that convert pressure now use siunit_pressure()
- Added Documentation/Amended documentation
- Rewritten Travis integration tests, now use siunit_pressure() instead of siunit()

MadMedicineMan committed:
Merge branch 'canister-display-fix-subprocs' into canister-capacity-fix

Mothblocks reviewed (approved):
Unit tests are good.
